### ISSUE TITLE
Wrong page in cache for customer_group

### ISSUE THREAD

**Issue by alemontalti** [nexcess/magento-turpentine#1051]
Hello,
I'm using Varnish in production for my ecommerce and I have really good performances on loading time and caching, but last week we should introduce a customer_group special pricing and we found a very bad issue: some page cached for "customer group" = X was served to standard user (no customer group cookie), therefore standard user could see special prices on catalog ecc...

(X = customer group id)

The problem could be reproduced this way:
- varnish full reset
- login for user with "customer_group = X"
- surfing a page: prices are good for user with "customer_group = X"
- another browser/session, surfing the same page WITHOUT login (standard user): I see the page early cached for "customer_group = X", so wrong prices 

My system configuration is below:
- PHP 5.4
- Varnish 3.0.7-1
- Turpentine 0.6.7 (supporting customer_group in hashing)

I have a staging environment to test the issue: same environment, same behavior. 

We're experiencing some problem with our client because of that issue, hope you can help us.

Best regards,

-
Ing. Alessandro Montalti
alessandro@cosmobile.net

**Comment by miguelbalparda** [comment]
I think you are caching your blocks as public instead of private, and that is why your blocks are being shown for all the users. Have you tried setting <access>private</access> to the conflicting method?

**Comment by davidverholen** [comment]
same problem here.

Caching with access set to private does work, but we want the block be cached per customer_group.

apperently now it seems to work for the product list, but still not for the product view. Here is my configuration (not sure if I need everything):

```xml
<?xml version="1.0"?>
<layout version="0.1.0">
    <catalog_category_layered>
        <reference name="product_list">
            <action method="setEsiOptions">
                <params>
                    <access>customer_group</access>
                    <registry_keys>
                        <current_category/>
                        <current_layer/>
                        <current_customer/>
                        <current_group/>
                    </registry_keys>
                    <flush_events>
                        <customer_logout/>
                        <customer_login/>
                    </flush_events>
                </params>
            </action>
        </reference>
    </catalog_category_layered>
    <catalog_product_view>
        <reference name="content">
            <action method="setEsiOptions">
                <params>
                    <access>customer_group</access>
                    <registry_keys>
                        <current_category/>
                        <current_product/>
                        <current_customer/>
                        <current_group/>
                    </registry_keys>
                    <flush_events>
                        <customer_logout/>
                        <customer_login/>
                    </flush_events>
                </params>
            </action>
        </reference>
    </catalog_product_view>
</layout>
```

**Comment by puecher** [comment]
Hi @davidverholen,

did you find a solution for catalog_product_view?

**Comment by davidverholen** [comment]
@puecher unfortunately still no solution for this, but haven't tried for a while

**Comment by ADDISON74** [comment]
I am looking for a solution too since we created groups for a customer. I confirm Turpentine needs customization for this situation.

**Comment by Stasevi4** [comment]
Any updates about customer groups ?

**Comment by aricwatson** [comment]
@davidverholen Sorry for the late reply

> apparently now it seems to work for the product list, but still not for the product view. Here is my configuration (not sure if I need everything):

The way to test this is to turn on the "Enable Debug Info" option in the Turpentine configuration options, then clear your varnish caches and reload a page that has a block which you're trying to cache separately by customer_group

You should see big HTML comments that start with `<!-- ESI URL: ` - look through that for the string `access/customer_group` - which means that the customer_group is correctly being set for that block. 

You can also test this via varnishlog - here's my approach. Start it up with `varnishlog > tmp.txt`, and then reload the page. Once the page is done loading, hit ctrl-c to stop varnishlog and look in tmp.txt file for something like this:

```
10 Hash         c /turpentine/esi/getBlock/method/esi/access/customer_group/ttl/3600/hmac/19f1c70ef0d0a9bbee1c0d831f61db045dcb33c283226b353a9685805969bd32/data/9NVhChT4naecfcRq.cdK2x6fJPX9t0ehvV.hI10YPFljM.NJpRtCFIjFG8UfGP1i07b2.5F4kK0vENhpQOSJOksFbnrM-CDFRMYElWTsmk5v1eTT
10 Hash         c www.turpentine-master.devsupport.nexcess.net
10 Hash         c
10 Hash         c gzip
10 Hash         c tlkjklwekll
```

This is showing you the items that are being added to the hash by which this ESI block will be stored and recalled in Varnish. The "tlkjklwekll" should be the user's 'customer_group' cookie value.

Turpentine (currently) only supports caching ESI blocks this way - not whole pages.

**Comment by Stasevi4** [comment]
I think i fix it by changing config file version-3.vcl
```
 if (req.http.X-Varnish-Esi-Access == "customer_group" &&
            req.http.Cookie ~ "customer_group=") {
        hash_data(regsub(req.http.Cookie, "^.*?customer_group=([^;]*);*.*$", "\1"));
    }
```
to 
```
 if (req.http.Cookie ~ "customer_group=") {
        hash_data(regsub(req.http.Cookie, "^.*?customer_group=([^;]*);*.*$", "\1"));
    }
```

**Comment by aricwatson** [comment]
OK - be aware that will cause the 'customer_group' (if the cookie is present) to be added to all cache objects (including pages and all ESI blocks).  

This may have some side effects - for example, if someone arrives on a page without a customer group cookie they'll be hitting one page cache object. If they someone arrives with a customer group cookie -t hey will hit a *different* page cache object. So you're fragmenting the page cache - which may be ok, but might lead to a lower hitrate.

**Comment by Stasevi4** [comment]
Yes we have about 10 customers group and its will create different cache object per group. How we can void this logic because we have only 1 catalog rule (affected price) in magento for wholesale group   others group have same content. Can we create additional configuration  in admin area to void creation of page cache object for customers group ? Example we need only 2 copy of cache pages 
first copy  for guest users and 9 group of registered customers  second for wholesales group they have  discount on price. 
Thank you

**Comment by aricwatson** [comment]
I'd suggest replacing this:

```

    if (req.http.X-Varnish-Esi-Access == "customer_group" &&
            req.http.Cookie ~ "customer_group=") {
        hash_data(regsub(req.http.Cookie, "^.*?customer_group=([^;]*);*.*$", "\1"));
    }

```

With this:

```

    if (req.http.Cookie ~ "customer_group=") {
       hash_data('customer_group');
    }

```

Which should just add 'customer_group' to the hash (if it's present in the cookie) when calculating it - all the different customer_groups would share this same hash, while users without that cookie will have their own hash. Please test, I haven't verified this!

**Comment by aricwatson** [comment]
That sounds like a good plan!

**Comment by ADDISON74** [comment]
It will be great if you find a fix. I do have a website where the owner is using groups for giving a discount level based on the order amounts. I found some issues with Varnish and now it is deactivated for a while. Hope you guys find a proper solution.

**Comment by aricwatson** [comment]
@ADDISON74 Have you tried caching the places where the price is displayed in an ESI block, with customer_group? Private might even work.

**Comment by Stasevi4** [comment]
I think it  will be nightmare when you have custom template to mark them private for example we use bestsellers side block, most viewed products, all block include product price and price must be different for customer_group.

**Comment by ADDISON74** [comment]
@aricwatson. i will try your suggestion but if I remember I already did with no success. Maybe I did something wrong or it wasn't the right block. I still keep an eye on this post if someone.

**Comment by aricwatson** [comment]
@Stasevi4 were you able to get it working?

**Comment by Stasevi4** [comment]
@aricwatson I create PR

**Comment by aricwatson** [comment]
This will only work with corresponding VCL change [mentioned above](https://github.com/nexcess/magento-turpentine/issues/1051#issuecomment-228127996) correct?

**Comment by Stasevi4** [comment]
It will create new customer_group if its will be selected in admin area and i just use in our production this config 
``` 
 if (req.http.Cookie ~ "customer_group=") {
       hash_data('customer_group');
    }
```
